File: app/api/signals.py
from fastapi import APIRouter

router = APIRouter(prefix="/signals", tags=["signals"])


@router.get("/")
async def list_signals(limit: int = 20, outcome: str | None = None, direction: str | None = None):
    from app.main import _app_state
    from app.signals.repository import SignalRepository
    async with _app_state.session_factory() as session:
        repo = SignalRepository(session)
        signals = await repo.get_signals_filtered(limit=limit, outcome=outcome, direction=direction)
    return {"data": signals, "error": None}


@router.get("/stats")
async def signal_stats():
    from app.main import _app_state
    from app.signals.repository import SignalRepository
    async with _app_state.session_factory() as session:
        repo = SignalRepository(session)
        stats = await repo.get_overall_stats()
    return {"data": stats, "error": None}


@router.get("/{signal_id}")
async def get_signal(signal_id: int):
    from app.main import _app_state
    from app.signals.repository import SignalRepository, _orm_to_dict
    async with _app_state.session_factory() as session:
        repo = SignalRepository(session)
        sig = await repo.get_signal_by_id(signal_id)
    if not sig:
        return {"data": None, "error": "Signal not found"}
    return {"data": _orm_to_dict(sig), "error": None}

File: frontend/src/api/endpoints/signals.ts
import { api } from '../client'
import type { Signal, SignalStats, SignalDirection, SignalOutcome } from '../types'

export const signalsApi = {
  list: (params?: { limit?: number; outcome?: SignalOutcome; direction?: SignalDirection }) => {
    const q = new URLSearchParams()
    if (params?.limit) q.set('limit', String(params.limit))
    if (params?.outcome) q.set('outcome', params.outcome)
    if (params?.direction) q.set('direction', params.direction)
    return api.get<Signal[]>(`/signals/?${q}`)
  },
  getStats: () => api.get<SignalStats>('/signals/stats'),
  get: (id: number) => api.get<Signal>(`/signals/${id}`),
}
